        <!DOCTYPE html>
        <html lang="en">
        <head><title>Exceptions in BPF [LWN.net]</title>
        <meta name="viewport" content="width=device-width, initial-scale=1">
<meta HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=utf-8">
<META NAME="robots" CONTENT="noai, noimageai">
        <link rel="icon" href="https://static.lwn.net/images/favicon.png"
              type="image/png">
        <link rel="alternate" type="application/rss+xml" title="LWN.net headlines" href="https://lwn.net/headlines/rss">
<link rel="alternate" type="application/rss+xml" title="Comments posted to this article" href="https://lwn.net/headlines/938435/">
        <link rel="stylesheet" href="/CSS/lwn">
<link rel="stylesheet" href="/CSS/nosub">

        
<script type="text/javascript">var p="http",d="static";if(document.location.protocol=="https:"){p+="s";d="engine";}var z=document.createElement("script");z.type="text/javascript";z.async=true;z.src=p+"://"+d+".adzerk.net/ados.js";var s=document.getElementsByTagName("script")[0];s.parentNode.insertBefore(z,s);</script>
<script type="text/javascript">
var ados_keywords = ados_keywords || [];
if( location.protocol=='https:' ) {
        ados_keywords.push('T:SSL');
} else {
        ados_keywords.push('T:HTTP');
}

var ados = ados || {};
ados.run = ados.run || [];
ados.run.push(function() {

ados_add_placement(4669, 20979, "azk13321_leaderboard", 4).setZone(16026);

ados_add_placement(4669, 20979, "azk93271_right_zone", [5,10,6]).setZone(16027);

ados_add_placement(4669, 20979, "azk31017_tracking", 20).setZone(20995);



ados_setKeywords(ados_keywords.join(', ')); 
ados_load();
});</script>

        </head>
        <body>
        <a name="t"></a>
<div id="menu"><a href="/"><img src="https://static.lwn.net/images/logo/barepenguin-70.png" class="logo"
                 border="0" alt="LWN.net Logo">
           <span class="logo">LWN<br>.net</span>
           <span class="logobl">News from the source</span></a>
           <a href="/"><img src="https://static.lwn.net/images/lcorner-ss.png" class="sslogo"
                 border="0" alt="LWN"></a><div class="navmenu-container">
           <ul class="navmenu">
        <li><a class="navmenu" href="#t"><b>Content</b></a><ul><li><a href="/current/">Weekly Edition</a></li><li><a href="/Archives/">Archives</a></li><li><a href="/Search/">Search</a></li><li><a href="/Kernel/">Kernel</a></li><li><a href="/Security/">Security</a></li><li><a href="/Calendar/">Events calendar</a></li><li><a href="/Comments/unread">Unread comments</a></li><li><hr></li><li><a href="/op/FAQ.lwn">LWN FAQ</a></li><li><a href="/op/AuthorGuide.lwn">Write for us</a></li></ul></li>
<li><a class="navmenu" href="#t"><b>Edition</b></a><ul><li><a href="/Articles/938713/">Return to the Front page</a></li></ul></li>
</ul></div>
</div> <!-- menu -->
<div class="not-handset"
            	     style="margin-left: 10.5em; display: block;">
                   <div class="not-print"> <div id="azk13321_leaderboard"></div> </div>
                </div>
            <div class="topnav-container">
<div class="not-handset"><form action="https://lwn.net/Login/" method="post" name="loginform"
                 class="loginform">
        <label><b>User:</b> <input type="text" name="uname" value="" size="8" id="uc" /></label> 
		<label><b>Password:</b> <input type="password" name="pword" size="8" id="pc" /></label> <input type="hidden" name="target" value="/Articles/938435/" /> <input type="submit" name="submit" value="Log in" /></form> |
           <form action="https://lwn.net/subscribe/" method="post" class="loginform">
           <input type="submit" name="submit" value="Subscribe" />
           </form> |
           <form action="https://lwn.net/Login/newaccount" method="post" class="loginform">
           <input type="submit" name="submit" value="Register" />
           </form>
        </div>
               <div class="handset-only">
               <a href="/subscribe/"><b>Subscribe</b></a> /
               <a href="/Login/"><b>Log in</b></a> /
               <a href="/Login/newaccount"><b>New account</b></a>
               </div>
               </div><div class="maincolumn flexcol">
<div class="middlecolumn">
<div class="PageHeadline">
<h1>Exceptions in BPF</h1>
</div>
<div class="ArticleText">
<blockquote class="ad">
<b>Benefits for LWN subscribers</b>
<p>
The primary benefit from <a href="/Promo/nst-nag5/subscribe">subscribing to LWN</a>
       is helping to keep us publishing, but, beyond that, subscribers get
       immediate access to all site content and access to a number of extra
       site features.  Please sign up today!
</blockquote>
<div class="FeatureByline">
           By <b>Jonathan Corbet</b><br>July 21, 2023</br>
           </div>
The BPF virtual machine in the kernel has been steadily gaining new
features for years, many of which add capabilities that C programmers do
not ordinarily have.  So, from one point of view, it was only a matter of
time before BPF gained support for exceptions.  As it turns out, though,
this "exceptions" feature is aimed at a specific use case, and its use in
most programs will be truly exceptional.
<p>
Kumar Kartikeya Dwivedi posted <a
href="/ml/bpf/20230713023232.1411523-1-memxor@gmail.com/">the BPF
exceptions patch set</a> on July&nbsp;13.  The API presented to BPF
programs is simple, taking the form of two <a
href="/Articles/856005/">kfuncs</a>.  To raise an exception, a BPF program
can call:
<p>
<pre>
    void bpf_throw(u64 cookie);
</pre>
<p>
A call to <tt>bpf_throw()</tt> will cause the program's entire call stack
to be unwound, and the program will return to its caller with (by default)
a return status of zero; the <tt>cookie</tt> value is ignored.  There is no
way for a program to catch an exception called further down the call stack.
It is, however, possible to define a function to be called after the call
stack has been unwound, but before control is returned to the caller:
<p>
<pre>
    void bpf_set_exception_callback(int (*callback)(u64));
</pre>
<p>
The given <tt>callback()</tt> will be called once unwinding is complete,
and will be passed the <tt>cookie</tt> value given to <tt>bpf_throw()</tt>;
its return value will then be returned to the original caller of the BPF
program.  There can be only one <tt>bpf_set_exception_callback()</tt> call
in a program; once the callback is set, it cannot be changed.
<p>
One might thus be forgiven for thinking that this exception mechanism does
not look like it does in other languages supporting the feature, and that
<tt>bpf_throw()</tt> might better be spelled <tt>exit()</tt>.  It clearly
is not meant to allow BPF programs to catch and respond to unusual
situations.  The use case for exceptions, as it turns out, is different and
unique to BPF.
<p>
BPF programs must, famously, convince the kernel's verifier that they are
safe to run before they can be successfully loaded.  Doing so requires
handling every possible case — even cases that the programmer knows can
never happen, but which the verifier is less certain about.  So, for
example, a BPF function far down the call stack might have to check that an
integer value is within a given range, even though the developer knows
that it must be, because the verifier does not know that.  The check must
do something reasonable in response to an out-of-bounds value and, perhaps,
return a failure status all the way back up the call chain, all for a case
that can never happen.
<p>
And, as we all know, developers are never wrong about cases that can never
happen.
<p>
As Dwivedi <a
href="/ml/bpf/CAP01T77m9ArXiKLgKW-uttFikirC97VXP=hmXFkS8JrSDXRYJg@mail.gmail.com/">described</a>,
exceptions are intended to address this problem:
<p>
<blockquote class="bq">
	The primary requirement was for implementing assertions within a
	program, which when untrue still ensure that the program terminates
	safely. Typically this would require the user to handle the other
	case, freeing any resources, and returning from a possibly deep
	callchain back to the kernel. Testing a condition can be used to
	update the verifier's knowledge about a particular register.
</blockquote>
<p>
So, in other words, the real reason for exceptions is to provide a
mechanism by which the verifier can be informed of invariants that the
developer knows about while having an emergency exit mechanism for those
times when the developer is wrong.  There is <a
href="/ml/bpf/20230713023232.1411523-10-memxor@gmail.com/">a set of
assertion macros</a> provided to make this feature easily available in BPF
programs.  So, for example, a developer will be able to write:
<p>
<pre>
    bpf_assert_lt(foo, 256);
</pre>
<p>
This assertion will perform the indicated test and, should it fail, make a
call to <tt>bpf_throw()</tt>.  Meanwhile, the verifier will be able to use
the knowledge that <tt>foo</tt> is, indeed, less than&nbsp;256 as it
evaluates the subsequent code.
<p>
There is one notable problem still, as described in the changelog to <a
href="/ml/bpf/20230713023232.1411523-7-memxor@gmail.com/">this patch</a> in
the series:
<p>
<blockquote class="bq">
	For now, bpf_throw invocation fails when lingering resources or
	locks exist in that path of the program. In a future followup,
	bpf_throw will be extended to perform frame-by-frame unwinding to
	release lingering resources for each stack frame, removing this
	limitation.
</blockquote>
<p>
Given that the verifier is now counting on <tt>bpf_throw()</tt> to prevent
execution from proceeding past a failed assertion, this seems like a
significant limitation indeed.  It could probably be used by a sufficiently
malicious developer to convince the verifier to accept a program that does
something unpleasant.  That suggests that implementing the frame-by-frame
unwinding will be a prerequisite to getting this work merged.
<p>
Both BPF and Rust are intended to make kernel programming safer, but they
take a different approach to the problem.  A Rust program will, by default,
panic if any of a large number of things goes wrong.  BPF programs,
instead, are intended to be verified as simply lacking that sort of wrong
behavior before they are ever allowed to execute.  BPF exceptions can be
seen as an admission that the "prove correctness before loading" approach
has its limits, and that sometimes it is necessary to just throw up your
hands and bail out.<br clear="all"><table class="IndexEntries">
           <tr><th colspan=2>Index entries for this article</th></tr>
           <tr><td><a href="/Kernel/Index">Kernel</a></td><td><a href="/Kernel/Index#BPF">BPF</a></td></tr>
            <tr><td><a href="/Kernel/Index">Kernel</a></td><td><a href="/Kernel/Index#Releases-6.7">Releases/6.7</a></td></tr>
            </table><br clear="all">
<hr width="60%%" align="left">
            <form action="/Login/" method="post">
            <input type="hidden" name="target" value="/Articles/938435/" />
            <input type="submit" name="login" value="Log in" /> to post comments
            <p>
        
</div> <!-- ArticleText -->
<p><a name="Comments"></a>
<a name="CommAnchor938907"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 21, 2023 16:25 UTC (Fri)
                               by <b>Cyberax</b> (<b>&#x272D; supporter &#x272D;</b>, #52523)
                              [<a href="/Articles/938907/">Link</a>] (11 responses)
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
Guys, really? You are now adding freaking exceptions.<br>
<p>
This is not just "feature creep", it's a "feature runaway train at 100mph".<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938907/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    <a name="CommAnchor938909"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 21, 2023 16:55 UTC (Fri)
                               by <b>adobriyan</b> (subscriber, #30858)
                              [<a href="/Articles/938909/">Link</a>] 
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
Superficially kuint devs did exceptions first:<br>
<p>
struct kunit_try_catch {<br>
        /* private: internal use only. */<br>
        struct kunit *test;<br>
        struct completion *try_completion;<br>
        int try_result;<br>
        kunit_try_catch_func_t try;<br>
        kunit_try_catch_func_t catch;<br>
        void *context;<br>
};<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938909/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    </details>
<a name="CommAnchor938919"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 21, 2023 20:00 UTC (Fri)
                               by <b>Smon</b> (guest, #104795)
                              [<a href="/Articles/938919/">Link</a>] (2 responses)
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
Couldn't a compiler simulate this already with existing functionality?<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938919/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    <a name="CommAnchor938976"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 22, 2023 16:00 UTC (Sat)
                               by <b>NYKevin</b> (subscriber, #129325)
                              [<a href="/Articles/938976/">Link</a>] (1 responses)
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
Maybe. Sort of. With a lot of effort.<br>
<p>
The problem is that you need to return all the way up the stack, and therefore you need to return some kind of "we're bailing out" status code to indicate the problem to the caller. But you probably already have return values at many of those call sites, so now you need to transform those return values in some way, probably into an option type or tagged union (or some equivalent). In principle that should be possible, but I don't know if BPF makes it straightforward or performant.<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938976/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    <a name="CommAnchor939019"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 24, 2023 4:13 UTC (Mon)
                               by <b>Cyberax</b> (<b>&#x272D; supporter &#x272D;</b>, #52523)
                              [<a href="/Articles/939019/">Link</a>] 
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
One of the bragging points of BPF was "but it always returns a value!", and that's why it's apparently better than WASM.<br>
<p>
With the addition of exceptions, this guarantee is lost.<br>
<p>
Not that it mattered either way in practice, but still. BPF is now just adding features without even considering their impact on the overall BPF model.<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/939019/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    </details>
</details>
</details>
<a name="CommAnchor938963"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 22, 2023 8:42 UTC (Sat)
                               by <b>jezuch</b> (subscriber, #52988)
                              [<a href="/Articles/938963/">Link</a>] (6 responses)
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
Maybe this thould be renamed as "panic", because it's doing the same thing as panic in Rust. These are no generic exceptions.<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938963/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    <a name="CommAnchor938966"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 22, 2023 9:15 UTC (Sat)
                               by <b>softball</b> (subscriber, #160655)
                              [<a href="/Articles/938966/">Link</a>] (3 responses)
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
A core difference is: BPF looks to treat throws as success-by-default (rc 0), Rust panics are failure-by-default. Rust panics have to be caught explicitly. My experience with Rust is limited so far, but catching panics seems rare and unidiomatic: there's already featureful error handling available (Result type), so any panics are usually serious enough to not be handlable.<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938966/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    <a name="CommAnchor938993"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 23, 2023 7:35 UTC (Sun)
                               by <b>xi0n</b> (subscriber, #138144)
                              [<a href="/Articles/938993/">Link</a>] (2 responses)
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
Panics in idiomatic Rust are only caught to prevent them from crossing an FFI boundary, i.e. escaping a Rust callback that’s been called from C. Otherwise, programs can set a global panic handler if the standard one is inadequate, but that is only really used for embedded and other no_std environments.<br>
<p>
This saying, the mechanism proposed here is so close to Rust panics (and Go panics) that NOT calling it such will only lead to confusion, esp. when “exception” is such an overloaded term already.<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938993/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    <a name="CommAnchor939138"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 25, 2023 6:59 UTC (Tue)
                               by <b>taladar</b> (subscriber, #68407)
                              [<a href="/Articles/939138/">Link</a>] 
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
Setting a custom panic handler is also frequently used in situations where the panic should be logged before exiting.<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/939138/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    </details>
<a name="CommAnchor950280"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Nov 5, 2023 13:03 UTC (Sun)
                               by <b>ibukanov</b> (guest, #3942)
                              [<a href="/Articles/950280/">Link</a>] 
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
Go panics are exactly the exceptions. They can be caught at arbitrary points and can pass arbitrary values across the stack.<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/950280/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    </details>
</details>
</details>
<a name="CommAnchor938970"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 22, 2023 12:07 UTC (Sat)
                               by <b>dezgeg</b> (subscriber, #92243)
                              [<a href="/Articles/938970/">Link</a>] (1 responses)
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
Sounds it would be too confusing when 'Kernel panic' is already a thing<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938970/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    <a name="CommAnchor939003"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 23, 2023 18:48 UTC (Sun)
                               by <b>rqosa</b> (subscriber, #24136)
                              [<a href="/Articles/939003/">Link</a>] 
      </p>
      
      </div>
      </summary>
      <p>Maybe naming it "error" (or something like that) would be a reasonable choice that isn't too similar to those 2 other terms?</p>
<p>(Java uses the word "error" in a similar sense in <tt>java.lang.Error</tt>'s name — i.e. for "<font class="QuotedText">serious problems that a reasonable application should not try to catch</font>".)</p>








      
          <div class="CommentReplyButton">
            <form action="/Articles/939003/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    </details>
</details>
</details>
</details>
<a name="CommAnchor938943"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 21, 2023 23:07 UTC (Fri)
                               by <b>randomguy3</b> (subscriber, #71063)
                              [<a href="/Articles/938943/">Link</a>] (1 responses)
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
I'd argue that rust borrow checks and panics are conceptually very similar to bpf verification and exceptions. In both cases, you are attempting to eliminate certain classes of failures before running the program, and providing a mechanism to assert invariants and cleanly handle bugs.<br>
<p>
The difference is primarily scope: as a general purpose language, rust's borrow checker is weaker than bpf's verification (with a few escape hatches provided as well), and its use of asserts/panics more extensive.<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938943/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    <a name="CommAnchor938965"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 22, 2023 9:07 UTC (Sat)
                               by <b>softball</b> (subscriber, #160655)
                              [<a href="/Articles/938965/">Link</a>] 
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
A core aspect they have in common is to err on the side of false-positives. False-negatives would be disastrous, and perfect, 1:1 deduction isn't possible (halting problem), so the only reasonable thing to do is disallow certain known-good (to the developer) scenarios.<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938965/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    </details>
</details>
<a name="CommAnchor938948"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 22, 2023 1:30 UTC (Sat)
                               by <b>walters</b> (subscriber, #7396)
                              [<a href="/Articles/938948/">Link</a>] (2 responses)
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
<span class="QuotedText">&gt; A Rust program will, by default, panic if any of a large number of things goes wrong</span><br>
<p>
I don’t think so.  If you’re talking about Rust code using std certainty there can be a lot of implicit OOM panics.   But kernel Rust doesn’t use std. <br>
<p>
Idiomatic Rust avoids gratuitous unwrap invocations and array accesses, etc.  Iterators can often remove implicit bounds checks too.  Now, writing probably panic-free code is an active topic.  <br>
<p>
But I don’t think “if any of a large number of things” is really accurate, on balance.<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938948/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    <a name="CommAnchor938955"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 22, 2023 4:24 UTC (Sat)
                               by <b>wahern</b> (subscriber, #37304)
                              [<a href="/Articles/938955/">Link</a>] (1 responses)
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
<span class="QuotedText">&gt; If you’re talking about Rust code using std certainty there can be a lot of implicit OOM panics. But kernel Rust doesn’t use std.</span><br>
<p>
True, kernel Rust has its own standard library, but most of it seems to be littered with '#[cfg(not(no_global_oom_handling))]' just like the userspace std library.<br>
<p>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938955/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    <a name="CommAnchor938983"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 22, 2023 18:28 UTC (Sat)
                               by <b>walters</b> (subscriber, #7396)
                              [<a href="/Articles/938983/">Link</a>] 
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
<span class="QuotedText">&gt; True, kernel Rust has its own standard library, but most of it seems to be littered with '#[cfg(not(no_global_oom_handling))]' just like the userspace std library.</span><br>
<p>
(We're now a bit past my relatively superficial knowledge of Linux kernel Rust but...)<br>
<p>
I'm pretty sure that's because they don't want a long term fork of alloc.rs and vec.rs etc.  The upstream Rust project has this config option I'm pretty sure *precisely* to help enable usage of the upstream battle-tested collections while disabling the APIs that will implicitly panic on OOM.<br>
<p>
And the Linux build system does pass  "--cfg no_global_oom_handling".  And the example Rust code does use try_push() not push(), etc.<br>
<p>
Or to say this simply: kernel Rust does not have implicit panics on OOM.<br>
<p>
What would certainly be interesting to try to evaluate is how many possible panics there are in any nontrivial kernel Rust code.  The classic example is somevec[offset].  I'd expect the number is greater than zero.  But are there "a lot"?  Are there enough where it *actually* feels like "panic if any of a large number of things goes wrong" is true?  I'm doubtful.<br>
<p>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938983/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    </details>
</details>
</details>
<a name="CommAnchor938938"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Rust and static safety</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 22, 2023 1:32 UTC (Sat)
                               by <b>geofft</b> (subscriber, #59789)
                              [<a href="/Articles/938938/">Link</a>] (1 responses)
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
<span class="QuotedText">&gt; A Rust program will, by default, panic if any of a large number of things goes wrong. BPF programs, instead, are intended to be verified as simply lacking that sort of wrong behavior before they are ever allowed to execute.</span><br>
<p>
Well... kind of. There are indeed a large number of things that cause a Rust panic (which, for kernelspace Rust code, turns into an oops, not a kernel panic), such as indexing an array out of bounds. But there are also a large number of things that are verified by the Rust compiler, preventing wrong behavior before it can ever execute, and I think that's sort of the selling point of Rust!<br>
<p>
A good example is the venerable null pointer. It's not actually a pointer - it's someone using a pointer type to convey there is nothing to point to. In C, you can attempt to "dereference" a null pointer, which is a fundamentally meaningless operation that will lead to incorrect behavior. In Rust, pointer types are defined as non-null, and the standard Optional data type when wrapping a pointer ends up in memory just like a nullable C pointer, but the language prevents you from using it if the value is equal to zero. An Optional type cannot be used directly; you have to use an if or match statement that breaks down the two possibilities, Some actual value or None. (Or you can call a function that does so - such as the standard .unwrap() function which will generate a Rust panic if it's None.)<br>
<p>
In other words, at compile time, a Rust program can be verified as never dereferencing a null reference.<br>
<p>
This can be generalized in a few ways. Rust also ensures that all references are to valid data, which is what the feared "borrow checker" does: if the compiler can't be convinced that the pointed-to data is still around when you're using the pointer, it will fail to compile. You can also imagine data types that have more than just null as special cases, such as the kernel's ERR_PTR scheme, where small negative values are actually errnos. A function that returns a char * might actually return (char *)-ENOMEM, aka ERR_PTR(-ENOMEM), and expect callers to check IS_ERR on the pointer before using it, with bad consequences if they forget. In Rust this is better defined as a data type that can be either a pointer or an error code (and indeed rust/kernel/error.rs defines it this way): you can't misinterpret an error code as a pointer or vice versa. Currently this isn't stored all in one pointer-sized data type the way it is in C, but it will be soon.<br>
<p>
The big difference between Rust and BPF in this context is that Rust treats all of these tools as aids to the programmer, which can be bypassed if needed if you're doing something complex, and BPF treats these as hard requirements and simply refuses to let you do complex things. Rust is trying to eliminate common types of mistakes, but it's not intended to be used in a way where the output is more highly privileged than the input. So it is possible (and quite common for interop with C) to "unsafely" produce a reference from somewhere, effectively telling Rust, hey trust me on this one, this is a valid pointer even though I can't prove it to you. BPF, on the other hand, is all about allowing userspace to load programs into the kernel without the security risk of loading a real kernel module. So it can't have any bypass mechanisms. (And so it needs something like the mechanism in this article to say, I can't prove this assumption to you, so you can just evaluate whether it's true at runtime and bail out of executing the program if it isn't.)<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938938/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    <a name="CommAnchor938977"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Rust and static safety</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 22, 2023 16:19 UTC (Sat)
                               by <b>randomguy3</b> (subscriber, #71063)
                              [<a href="/Articles/938977/">Link</a>] 
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
this is kind of what I was trying to get at with my comment, but you did a much better job!<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938977/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    </details>
</details>
<a name="CommAnchor938989"></a>
    <details class="CommentBox" open>
      <summary><h3 class="CommentTitle">Exceptions in BPF</h3>
      <div class="AnnLine">
      <p class="CommentPoster"> Posted Jul 23, 2023 2:44 UTC (Sun)
                               by <b>iteratedlateralus</b> (guest, #102183)
                              [<a href="/Articles/938989/">Link</a>] 
      </p>
      
      </div>
      </summary>
      <div class="FormattedComment">
I think this is the natural progression of things technological. It becomes easier and easier to build abstractions on top of something that used to be a lot of hard work to accomplish even the most basic of tasks. I think the BPF feature set should be purposefully limited in scope due to the valuable attack surface that features create.<br>
</div>

      
          <div class="CommentReplyButton">
            <form action="/Articles/938989/comment" method="post">
            <input type="submit" value="Reply to this comment">
            </form>
          </div>
        
     <p>
     
    </details>
</div> <!-- middlecolumn -->
<div class="rightcol not-print">
<div id="azk93271_right_zone"></div>
</div>
</div> <!-- maincolumn -->

            <br clear="all">
            <center>
            <P>
            <span class="ReallySmall">
            Copyright &copy; 2023, Eklektix, Inc.<BR>
            This article may be redistributed under the terms of the
              <a href="http://creativecommons.org/licenses/by-sa/4.0/">Creative
              Commons CC BY-SA 4.0</a> license<br>
            Comments and public postings are copyrighted by their creators.<br>
            Linux  is a registered trademark of Linus Torvalds<br>
            </span>
            </center>
            
            </body></html>
